File: microservices/model-categorizer/classifiers/classifier.go
package classifiers

import (
	"strconv"
	"strings"
	"time"
)

// Constants for model organization
const (
	// Providers
	ProviderOpenAI     = "openai"
	ProviderAnthropicA = "anthropic"
	ProviderGemini     = "gemini"
	ProviderMeta       = "meta"
	ProviderMistral    = "mistral"
	ProviderOther      = "other"

	// Series
	SeriesClaude3 = "Claude 3"
	SeriesClaude2 = "Claude 2"
	SeriesClaude1 = "Claude 1"

	// OpenAI Types
	TypeO    = "O Series"
	Type35   = "GPT 3.5"
	Type4    = "GPT 4"
	Type45   = "GPT 4.5"
	TypeMini = "Mini"

	// Other Types
	TypeOpus      = "Opus"
	TypeSonnet    = "Sonnet"
	TypeHaiku     = "Haiku"
	TypeThinking  = "Thinking"
	TypePro       = "Pro"
	TypeGemma     = "Gemma"
	TypeFlashLite = "Flash Lite"
	TypeFlash     = "Flash"
	TypeVision    = "Vision"
	TypeStandard  = "Standard"
	TypeEmbedding = "Embedding"
	TypeImage     = "Image Generation"

	// Version constants for improved consistency
	Version10 = "1.0"
	Version15 = "1.5"
	Version20 = "2.0"
	Version25 = "2.5"
	Version30 = "3.0"
	Version35 = "3.5"
	Version37 = "3.7"
	Version40 = "4.0"
	Version45 = "4.5"

	// Capabilities
	CapVision          = "vision"
	CapFunctionCalling = "function-calling"
	CapEmbedding       = "embedding"
	CapChat            = "chat"
)

// ModelMetadata contains organized model information
type ModelMetadata struct {
	Provider       string
	Series         string
	Type           string
	Variant        string
	Context        int
	Capabilities   []string
	IsMultimodal   bool
	IsExperimental bool
	DisplayName    string
}

// ModelClassifier helps efficiently classify models
type ModelClassifier struct {
	patterns *PatternMatcher
	context  *ContextResolver
	defaults *DefaultModels
}

// NewModelClassifier creates a new model classifier with improved hierarchical patterns
func NewModelClassifier() *ModelClassifier {
	return &ModelClassifier{
		patterns: NewPatternMatcher(),
		context:  NewContextResolver(),
		defaults: NewDefaultModels(),
	}
}

// ClassifyModel takes a model name and returns a structured metadata object
func (mc *ModelClassifier) ClassifyModel(modelName, providerHint string) ModelMetadata {
	origName := modelName
	modelLower := strings.ToLower(modelName)
	var metadata ModelMetadata
	if mc.isImageGenerationModel(modelLower) {
		metadata = mc.createImageGenerationMetadata(modelLower, providerHint)
	} else if mc.isEmbeddingModel(modelLower) {
		metadata = mc.createEmbeddingModelMetadata(modelLower, providerHint)
	} else {
		metadata = mc.buildStandardModelMetadata(modelLower, providerHint)
	}
	metadata.DisplayName = strings.ReplaceAll(origName, "-", " ")
	return metadata
}

// createImageGenerationMetadata creates metadata for image generation models
func (mc *ModelClassifier) createImageGenerationMetadata(modelName, providerHint string) ModelMetadata {
	return ModelMetadata{
		Provider:     mc.determineProvider(modelName, providerHint),
		Series:       TypeImage,
		Type:         TypeImage,
		Variant:      "Image Generation",
		Capabilities: []string{TypeImage},
		IsMultimodal: false,
	}
}

// createEmbeddingModelMetadata creates metadata for embedding models
func (mc *ModelClassifier) createEmbeddingModelMetadata(modelName, providerHint string) ModelMetadata {
	return ModelMetadata{
		Provider:     mc.determineProvider(modelName, providerHint),
		Series:       TypeEmbedding,
		Type:         TypeEmbedding,
		Variant:      "Embedding",
		Capabilities: []string{CapEmbedding},
		IsMultimodal: false,
	}
}

// buildStandardModelMetadata builds metadata for standard LLM models
func (mc *ModelClassifier) buildStandardModelMetadata(modelName, providerHint string) ModelMetadata {
	// Start with empty metadata
	metadata := ModelMetadata{
		Provider:     mc.determineProvider(modelName, providerHint),
		Capabilities: []string{},
	}

	// Determine series based on provider
	metadata.Series = mc.determineSeries(modelName, metadata.Provider)

	// Determine type based on provider and series
	metadata.Type = mc.determineType(modelName, metadata.Provider, metadata.Series)

	// Determine variant (version)
	metadata.Variant = mc.determineVariant(modelName, metadata.Provider, metadata.Series)

	// Determine context size
	metadata.Context = mc.getContextSize(modelName)

	// Determine capabilities
	metadata.Capabilities = mc.detectCapabilities(modelName, metadata.Provider, metadata.Series)

	// Set multimodal flag
	metadata.IsMultimodal = mc.isMultimodal(modelName, metadata.Capabilities, metadata.Series)

	// Set experimental flag
	metadata.IsExperimental = mc.isExperimental(modelName)

	return metadata
}

// determineProvider identifies the model provider from name
func (mc *ModelClassifier) determineProvider(modelName, providerHint string) string {
	// Check provider hint first if provided
	if providerHint != "" {
		providerLower := strings.ToLower(providerHint)
		if provider := mc.patterns.matchProviderByName(providerLower); provider != "" {
			return provider
		}
	}

	// Handle OpenRouter prefix: "provider/model"
	if strings.Contains(modelName, "/") {
		parts := strings.SplitN(modelName, "/", 2)
		potentialProvider := strings.ToLower(parts[0])
		if provider := mc.patterns.matchProviderByName(potentialProvider); provider != "" {
			return provider
		}
	}

	// Match provider by patterns
	if provider := mc.patterns.matchProviderByPattern(modelName); provider != "" {
		return provider
	}

	// Default provider if none matched
	return ProviderOther
}

// determineSeries identifies the model series based on name and provider
func (mc *ModelClassifier) determineSeries(modelName, provider string) string {
	// Provider-specific series determination
	switch provider {
	case ProviderOpenAI:
		if modelName[0] == 'o' {
			return "O"
		}
		if modelName[0] == 'g' {
			return "GPT"
		}
		if modelName[0] == 'd' {
			return "DALL-E"
		}
	case ProviderAnthropicA:
		if series := mc.patterns.matchClaudeVersion(modelName); series != "" {
			return series
		}

	case ProviderGemini:
		return mc.patterns.matchGeminiVersion(modelName)
	}

	// Generic fallback series detection
	if series := mc.patterns.matchSeriesByPattern(modelName); series != "" {
		return series
	}

	// Default series if none matched
	return "General"
}

// determineType identifies the model type based on name, provider and series
func (mc *ModelClassifier) determineType(modelName, provider, series string) string {
	modelLower := strings.ToLower(modelName)

	// Provider-specific type determination
	switch provider {
	case ProviderOpenAI:
		return mc.patterns.matchOpenAIType(modelLower)

	case ProviderAnthropicA:
		return mc.patterns.matchAnthropicType(modelLower)

	case ProviderGemini:
		return mc.patterns.matchGeminiType(modelLower)
	}

	// Generic type detection based on patterns
	if type_ := mc.patterns.matchTypeByPattern(modelLower); type_ != "" {
		return type_
	}

	// Default type if none matched
	return TypeStandard
}

// determineVariant extracts specific version information
func (mc *ModelClassifier) determineVariant(modelName, provider, series string) string {
	modelLower := strings.ToLower(modelName)

	// Provider-specific variant detection
	switch provider {
	case ProviderOpenAI:
		if variant := mc.patterns.matchOpenAIVariant(modelLower); variant != "" {
			return variant
		}

	case ProviderAnthropicA:
		if variant := mc.patterns.matchAnthropicVariant(modelLower); variant != "" {
			return variant
		}

	case ProviderGemini:
		if variant := mc.patterns.buildGeminiVariant(modelLower); variant != "" {
			return variant
		}
	}

	// If we couldn't determine a specific variant, try to extract version info
	if variant := extractVersionVariant(modelName, series); variant != "" {
		return variant
	}

	// Default to series name if no specific variant is found
	return series
}

// detectCapabilities identifies model capabilities from the model name
func (mc *ModelClassifier) detectCapabilities(modelName, provider, series string) []string {
	capabilities := make(map[string]bool)
	modelLower := strings.ToLower(modelName)

	// Get model type for provider-specific rules
	modelType := mc.determineType(modelLower, provider, series)

	// Add capabilities based on model traits
	mc.patterns.addCapabilities(capabilities, modelType, modelLower, provider, series)

	// Chat capability for all models (default)
	capabilities[CapChat] = true

	// Convert map to slice
	result := make([]string, 0, len(capabilities))
	for cap := range capabilities {
		result = append(result, cap)
	}

	return result
}

// isEmbeddingModel checks if a model is for embeddings
func (mc *ModelClassifier) isEmbeddingModel(modelName string) bool {
	modelLower := strings.ToLower(modelName)
	return strings.Contains(modelLower, "embedding") ||
		strings.Contains(modelLower, "embed") ||
		strings.Contains(modelLower, "text-embedding")
}

// isImageGenerationModel checks if a model is for image generation
func (mc *ModelClassifier) isImageGenerationModel(modelName string) bool {
	modelLower := strings.ToLower(modelName)
	return strings.Contains(modelLower, "dall-e") ||
		strings.Contains(modelLower, "image") ||
		strings.Contains(modelLower, "midjourney") ||
		strings.Contains(modelLower, "stable-diffusion")
}

// isMultimodal determines if a model has multimodal capabilities
func (mc *ModelClassifier) isMultimodal(modelName string, capabilities []string, series string) bool {
	// Check in capabilities
	for _, cap := range capabilities {
		if cap == CapVision {
			return true
		}
	}

	// Extract model type from name to handle models
	modelLower := strings.ToLower(modelName)
	modelType := mc.determineType(modelLower, mc.determineProvider(modelLower, ""), series)

	// Check for multimodal capabilities based on type and series
	if modelType == Type4 || modelType == Type45 || modelType == TypeO ||
		series == SeriesClaude3 || strings.Contains(series, "Gemini") {
		return true
	}

	// Check model name
	return strings.Contains(modelLower, "vision") ||
		strings.Contains(modelLower, "multimodal")
}

// isExperimental checks if a model is experimental
func (mc *ModelClassifier) isExperimental(modelName string) bool {
	modelLower := strings.ToLower(modelName)
	return strings.Contains(modelLower, "experimental") ||
		strings.Contains(modelLower, "preview") ||
		strings.Contains(modelLower, "alpha") ||
		strings.Contains(modelLower, "beta")
}

// IsDefaultModelName checks if a model is a default version
func (mc *ModelClassifier) IsDefaultModelName(modelName string) bool {
	return mc.defaults.isDefault(modelName) ||
		strings.Contains(strings.ToLower(modelName), "latest")
}

// getContextSize determines a model's context window based on its name
func (mc *ModelClassifier) getContextSize(modelName string) int {
	return mc.context.getContextSize(modelName)
}

// GetModelHierarchy returns hierarchy information (provider, series, type, variant)
func (mc *ModelClassifier) GetModelHierarchy(modelName string, provider string) (string, string, string, string) {
	metadata := mc.ClassifyModel(modelName, provider)
	return metadata.Provider, metadata.Series, metadata.Type, metadata.Variant
}

// GetSeriesAndVariant (maintained for backward compatibility)
func GetSeriesAndVariant(modelName string) (string, string) {
	classifier := NewModelClassifier()
	metadata := classifier.ClassifyModel(modelName, "")
	return metadata.Series, metadata.Variant
}

// NormalizeModelName removes provider prefixes from OpenRouter model IDs
func NormalizeModelName(modelID, provider string) string {
	// Handle OpenRouter models which often contain provider names
	if strings.ToLower(provider) == "openrouter" {
		// Remove provider prefixes like "anthropic/" or "openai/"
		parts := strings.SplitN(modelID, "/", 2)
		if len(parts) == 2 {
			// List of known providers
			knownProviders := []string{"anthropic", "openai", "google", "meta-llama", "mistralai"}
			subProvider := strings.ToLower(parts[0])

			for _, provider := range knownProviders {
				if subProvider == provider {
					return parts[1]
				}
			}
		}
	}
	return modelID
}

// extractVersionVariant extracts version info from a model name
func extractVersionVariant(modelName, series string) string {
	versionNumbers := ExtractVersionNumbers(modelName)

	if len(versionNumbers) > 0 {
		versionStr := strings.Join(strings.Fields(strings.Map(func(r rune) rune {
			if r >= '0' && r <= '9' || r == '.' {
				return r
			}
			return ' '
		}, modelName)), ".")

		if versionStr != "" {
			return series + " " + versionStr
		}
	}

	return ""
}

// ExtractVersionNumbers extracts numeric parts from version strings
func ExtractVersionNumbers(version string) []int {
	var numbers []int
	var currentNumber string

	for _, c := range version {
		if c >= '0' && c <= '9' {
			currentNumber += string(c)
		} else if currentNumber != "" {
			if num, err := strconv.Atoi(currentNumber); err == nil {
				numbers = append(numbers, num)
			}
			currentNumber = ""
		}
	}

	// Add the last number
	if currentNumber != "" {
		if num, err := strconv.Atoi(currentNumber); err == nil {
			numbers = append(numbers, num)
		}
	}

	return numbers
}

// IsNewerVersion compares version strings to determine if a is newer than b
func IsNewerVersion(a, b string) bool {
	// Handle date-based version strings (like "20240307")
	aDate, aErr := time.Parse("20060102", a)
	bDate, bErr := time.Parse("20060102", b)

	// If both are dates, compare them
	if aErr == nil && bErr == nil {
		return aDate.After(bDate)
	}

	// Extract numeric version parts if present
	aParts := ExtractVersionNumbers(a)
	bParts := ExtractVersionNumbers(b)

	// If both have numeric parts, compare them
	if len(aParts) > 0 && len(bParts) > 0 {
		minLen := len(aParts)
		if len(bParts) < minLen {
			minLen = len(bParts)
		}

		for i := 0; i < minLen; i++ {
			if aParts[i] != bParts[i] {
				return aParts[i] > bParts[i]
			}
		}

		// If all common parts are equal, longer is newer
		return len(aParts) > len(bParts)
	}

	// If one is a date and the other isn't, date is newer
	if aErr == nil {
		return true
	}
	if bErr == nil {
		return false
	}

	// Fallback to string comparison
	return a > b
}

// GetStandardizedVersion returns a standardized version string from a model name
func (mc *ModelClassifier) GetStandardizedVersion(modelName string) string {
	// Convert to lowercase for consistent matching
	modelLower := strings.ToLower(modelName)

	// Check for specific version numbers
	if strings.Contains(modelLower, "4.5") {
		return Version45
	} else if strings.Contains(modelLower, "4.0") || strings.Contains(modelLower, "4o") {
		return Version40
	} else if strings.Contains(modelLower, "3.7") {
		return Version37
	} else if strings.Contains(modelLower, "3.5") {
		return Version35
	} else if strings.Contains(modelLower, "3.0") {
		return Version30
	} else if strings.Contains(modelLower, "2.5") {
		return Version25
	} else if strings.Contains(modelLower, "2.0") {
		return Version20
	} else if strings.Contains(modelLower, "1.5") {
		return Version15
	} else if strings.Contains(modelLower, "1.0") {
		return Version10
	}

	// If no version is identified, return empty string
	return ""
}

File: Deprecated/Microservice/model-categorizer/node/classifiers/classifier.js
const {
  ProviderOpenAI,
  ProviderAnthropicA,
  ProviderGemini,
  ProviderOther,
  ProviderOpenrouter,
  SeriesClaude3,
  TypeO,
  Type35,
  Type4,
  Type45,
  TypeStandard,
  TypeEmbedding,
  TypeImage,
  CapVision,
} = require("./constants");
const { PatternMatcher } = require("./patternMatcher");
const { ContextResolver } = require("./contextResolver");
const { DefaultModels } = require("./defaultModels");

class ModelClassifier {
  constructor() {
    this.patterns = new PatternMatcher();
    this.context = new ContextResolver();
    this.defaults = new DefaultModels();
  }

  ClassifyModel(modelID, providerHint) {
    const modelLower = (modelID || "").toLowerCase();
    let metadata;
    if (this.isImageGenerationModel(modelLower)) {
      metadata = this.createImageGenerationMetadata(modelLower, providerHint);
    } else if (this.isEmbeddingModel(modelLower)) {
      metadata = this.createEmbeddingModelMetadata(modelLower, providerHint);
    } else {
      metadata = this.buildStandardModelMetadata(modelLower, providerHint);
    }
    return metadata;
  }

  createImageGenerationMetadata(modelName, providerHint) {
    return {
      Provider: this.determineProvider(modelName, providerHint),
      Series: TypeImage,
      Type: TypeImage,
      Variant: "Image Generation",
      Capabilities: [TypeImage],
      IsMultimodal: false,
    };
  }

  createEmbeddingModelMetadata(modelName, providerHint) {
    return {
      Provider: this.determineProvider(modelName, providerHint),
      Series: TypeEmbedding,
      Type: TypeEmbedding,
      Variant: "Embedding",
      Capabilities: ["embedding"],
      IsMultimodal: false,
    };
  }

  buildStandardModelMetadata(modelName, providerHint) {
    const metadata = {
      Provider: this.determineProvider(modelName, providerHint),
      Capabilities: [],
    };

    metadata.Series = this.determineSeries(modelName, metadata.Provider);
    metadata.Type = this.determineType(modelName, metadata.Provider, metadata.Series);
    metadata.Variant = this.determineVariant(modelName, metadata.Provider, metadata.Series);
    metadata.Context = this.GetContextSize(modelName);
    metadata.Capabilities = this.detectCapabilities(modelName, metadata.Provider, metadata.Series);
    metadata.IsMultimodal = this.isMultimodal(modelName, metadata.Capabilities, metadata.Series);
    metadata.IsExperimental = this.isExperimental(modelName);
    return metadata;
  }

  determineProvider(modelName, providerHint) {
    if (providerHint) {
      const providerLower = providerHint.toLowerCase();
      const provider = this.patterns.matchProviderByName(providerLower);
      if (provider) return provider;
    }
    if ((modelName || "").includes("/")) {
      const parts = modelName.split("/", 2);
      const potentialProvider = (parts[0] || "").toLowerCase();
      const provider = this.patterns.matchProviderByName(potentialProvider);
      if (provider) return provider;
    }
    const provider = this.patterns.matchProviderByPattern(modelName);
    if (provider) return provider;
    return ProviderOther;
  }

  determineSeries(modelName, provider) {
    switch (provider) {
      case ProviderOpenAI:
        if (modelName[0] === "o") return "O";
        if (modelName[0] === "g") return "GPT";
        if (modelName[0] === "d") return "DALL-E";
        break;
      case ProviderAnthropicA:
        {
          const series = this.patterns.matchClaudeVersion(modelName);
          if (series) return series;
        }
        break;
      case ProviderGemini:
        return this.patterns.matchGeminiVersion(modelName);
    }
    const series = this.patterns.matchSeriesByPattern(modelName);
    if (series) return series;
    return "General";
  }

  determineType(modelName, provider, series) {
    const m = modelName.toLowerCase();
    switch (provider) {
      case ProviderOpenAI:
        return this.patterns.matchOpenAIType(m);
      case ProviderAnthropicA:
        return this.patterns.matchAnthropicType(m);
      case ProviderGemini:
        return this.patterns.matchGeminiType(m);
    }
    const t = this.patterns.matchTypeByPattern(m);
    if (t) return t;
    return TypeStandard;
  }

  determineVariant(modelName, provider, series) {
    const m = modelName.toLowerCase();
    switch (provider) {
      case ProviderOpenAI: {
        const v = this.patterns.matchOpenAIVariant(m);
        if (v) return v;
        break;
      }
      case ProviderAnthropicA: {
        const v = this.patterns.matchAnthropicVariant(m);
        if (v) return v;
        break;
      }
      case ProviderGemini: {
        const v = this.patterns.buildGeminiVariant(m);
        if (v) return v;
        break;
      }
    }
    const variant = extractVersionVariant(modelName, series);
    if (variant) return variant;
    return series;
  }

  detectCapabilities(modelName, provider, series) {
    const caps = {};
    const m = modelName.toLowerCase();
    const modelType = this.determineType(m, provider, series);
    this.patterns.addCapabilities(caps, modelType, m, provider, series);
    caps["chat"] = true;
    const result = Object.keys(caps);
    result.sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()));
    return result;
  }

  isEmbeddingModel(modelName) {
    const m = modelName.toLowerCase();
    return m.includes("embedding") || m.includes("embed") || m.includes("text-embedding");
  }

  isImageGenerationModel(modelName) {
    const m = modelName.toLowerCase();
    return m.includes("dall-e") || m.includes("image") || m.includes("midjourney") || m.includes("stable-diffusion");
  }

  isMultimodal(modelName, capabilities, series) {
    if (capabilities && capabilities.includes(CapVision)) return true;
    const m = modelName.toLowerCase();
    const type = this.determineType(m, this.determineProvider(m, ""), series);
    if (type === Type4 || type === Type45 || type === TypeO || series === SeriesClaude3 || (series && series.includes("Gemini"))) return true;
    return m.includes("vision") || m.includes("multimodal");
  }

  isExperimental(modelName) {
    const m = modelName.toLowerCase();
    return m.includes("experimental") || m.includes("preview") || m.includes("alpha") || m.includes("beta");
  }

  IsDefaultModelName(modelName) {
    return this.defaults.IsDefaultModel(modelName) || (modelName || "").toLowerCase().includes("latest");
  }

  GetContextSize(modelName) {
    return this.context.GetContextSize(modelName);
  }

  GetModelHierarchy(modelID, provider) {
    const metadata = this.ClassifyModel(modelID, provider);
    return [metadata.Provider, metadata.Series, metadata.Type, metadata.Variant];
  }

  GetStandardizedVersion(modelName) {
    const m = (modelName || "").toLowerCase();
    if (m.includes("4.5")) return "4.5";
    else if (m.includes("4.0") || m.includes("4o")) return "4.0";
    else if (m.includes("3.7")) return "3.7";
    else if (m.includes("3.5")) return "3.5";
    else if (m.includes("3.0")) return "3.0";
    else if (m.includes("2.5")) return "2.5";
    else if (m.includes("2.0")) return "2.0";
    else if (m.includes("1.5")) return "1.5";
    else if (m.includes("1.0")) return "1.0";
    return "";
  }
}

function NormalizeModelName(modelID, provider) {
  if ((provider || "").toLowerCase() === ProviderOpenrouter) {
    const parts = (modelID || "").split("/", 2);
    if (parts.length === 2) {
      const knownProviders = ["anthropic", "openai", "google", "gemini", "meta-llama", "mistralai"];
      const sub = (parts[0] || "").toLowerCase();
      if (knownProviders.includes(sub)) return parts[1];
    }
  }
  return modelID;
}

function extractVersionVariant(modelName, series) {
  const versionNumbers = ExtractVersionNumbers(modelName);
  if (versionNumbers.length > 0) {
    const versionStr = (modelName || "")
      .split("")
      .map((c) => (/[0-9.]/.test(c) ? c : " "))
      .join("")
      .trim()
      .split(/\s+/)
      .join(".");
    if (versionStr) return `${series} ${versionStr}`;
  }
  return "";
}

function ExtractVersionNumbers(version) {
  const numbers = [];
  let current = "";
  for (const c of version || "") {
    if (c >= "0" && c <= "9") current += c;
    else if (current) {
      numbers.push(parseInt(current, 10));
      current = "";
    }
  }
  if (current) numbers.push(parseInt(current, 10));
  return numbers;
}

function IsNewerVersion(a, b) {
  const dateRe = /^\d{8}$/;
  const parseDate = (s) => (dateRe.test(s) ? new Date(s.slice(0, 4) + "-" + s.slice(4, 6) + "-" + s.slice(6, 8)) : null);
  const aDate = parseDate(a);
  const bDate = parseDate(b);
  if (aDate && bDate) return aDate > bDate;
  const aParts = ExtractVersionNumbers(a);
  const bParts = ExtractVersionNumbers(b);
  if (aParts.length && bParts.length) {
    const minLen = Math.min(aParts.length, bParts.length);
    for (let i = 0; i < minLen; i++) {
      if (aParts[i] !== bParts[i]) return aParts[i] > bParts[i];
    }
    return aParts.length > bParts.length;
  }
  if (aDate) return true;
  if (bDate) return false;
  return a > b;
}

module.exports = {
  ModelClassifier,
  NormalizeModelName,
  ExtractVersionNumbers,
  IsNewerVersion,
};


